File: PythonCodes/Fig02_05.py
#%%
"""
Created on Thu Nov 28 2018
Generating Monte Carlo paths under P and Q measure
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GeneratePathsGBM(NoOfPaths,NoOfSteps,T,r,sigma,S_0):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    X = np.zeros([NoOfPaths, NoOfSteps+1])
    S = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
        
    X[:,0] = np.log(S_0)
    
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
     
        X[:,i+1] = X[:,i] + (r - 0.5 * sigma * sigma) * dt + sigma *\
        np.power(dt, 0.5)*Z[:,i]
        time[i+1] = time[i] +dt
        
    # Compute exponent of ABM

    S = np.exp(X)
    paths = {"time":time,"S":S}
    return paths
    
def MainCode():
    NoOfPaths = 8
    NoOfSteps = 1000
    S_0       = 1
    r         = 0.05
    mu        = 0.15
    sigma     = 0.1
    T         = 10

    # Money-savings account

    M         = lambda t: np.exp(r * t)
    
    # Monte Carlo paths

    pathsQ    = GeneratePathsGBM(NoOfPaths,NoOfSteps,T,r,sigma,S_0)
    S_Q       = pathsQ["S"]
    pathsP = GeneratePathsGBM(NoOfPaths,NoOfSteps,T,mu,sigma,S_0)
    S_P = pathsP["S"]
    time= pathsQ["time"]    
    
    # Discounted stock paths

    S_Qdisc = np.zeros([NoOfPaths,NoOfSteps+1])
    S_Pdisc = np.zeros([NoOfPaths,NoOfSteps+1])
    i = 0
    for i, ti in enumerate(time):
        S_Qdisc[:, i] = S_Q[:,i]/M(ti) 
        S_Pdisc[:, i] = S_P[:,i]/M(ti) 
    
    # S(T)/M(T) with stock growing with rate r

    plt.figure(1)
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("S(t)")
    eSM_Q = lambda t: S_0 * np.exp(r *t) / M(t)
    plt.plot(time,eSM_Q(time),'r--')
    plt.plot(time, np.transpose(S_Qdisc),'blue')   
    plt.legend(['E^Q[S(t)/M(t)]','paths S(t)/M(t)'])
    
    # S(T)/M(T) with stock growing with rate mu

    plt.figure(2)
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("S(t)")
    eSM_P = lambda t: S_0 * np.exp(mu *t) / M(t)
    plt.plot(time,eSM_P(time),'r--')
    plt.plot(time, np.transpose(S_Pdisc),'blue')   
    plt.legend(['E^P[S(t)/M(t)]','paths S(t)/M(t)'])
    
MainCode()

File: MatlabCodes/Chapter 02/Fig02_05.m
function PathsUnderQandPmeasure
close all;
randn('seed',3)
NoOfPaths = 8;
NoOfSteps = 1000;

T  = 10;
dt = T/NoOfSteps;

% Normals

Z1 = random('normal',0,1,[NoOfPaths,NoOfSteps])*sqrt(dt);
Z2 = random('normal',0,1,[NoOfPaths,NoOfSteps])*sqrt(dt);

% Parameters

S_0   = 1;
r     = 0.05;
mu    = 0.15;
sigma = 0.1;

% Generation of the paths

S1      = zeros(NoOfPaths,NoOfSteps);
S1(:,1) = S_0;
S2      = zeros(NoOfPaths,NoOfSteps);
S2(:,1) = S_0;

for i=1:NoOfSteps
    Z1(:,i) = (Z1(:,i)- mean(Z1(:,i)))/std(Z1(:,i))*sqrt(dt);
    Z2(:,i) = (Z2(:,i)- mean(Z2(:,i)))/std(Z2(:,i))*sqrt(dt);
    
    S1(:,i+1)=S1(:,i)+mu*S1(:,i)*dt + sigma*S1(:,i).*Z1(:,i);
    S2(:,i+1)=S2(:,i)+r*S2(:,i)*dt + sigma*S2(:,i).*Z2(:,i);
end

%% Plotting the results

time=0:dt:T;
figure(1)
hold on
plot(time,S_0*exp(mu*time)./exp(r*time),'--r','linewidth',2)
plot(time,S1.*exp(-r*time),'linewidth',1,'color',[0 0.45 0.74])
a=axis;
a(3)=0;
axis(a);
grid on;
xlabel('time')
lab=ylabel('$$\frac{S(t)}{M(t)}$$');
set(lab,'Interpreter','latex');
leg = legend('$$E^P\Big[\frac{S(t)}{M(t)}\Big]$$','paths $$\frac{S(t)}{M(t)}$$');
set(leg,'Interpreter','latex');

plot(time,S_0*exp(mu*time)./exp(r*time),'--r','linewidth',2)
titleT=title('Monte Carlo Paths of $$\frac{S(t)}{M(t)}$$ under real-world measure, P');
set(titleT,'Interpreter','latex');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

figure(2)
hold on
plot(time,S_0*ones(length(time),1),'--r','linewidth',2)
plot(time,S2.*exp(-r*time),'linewidth',1,'color',[0 0.45 0.74])

a(3)=0;
axis(a);
grid on;
xlabel('time')
lab=ylabel('$$\frac{S(t)}{M(t)}$$');
set(lab,'Interpreter','latex');
leg = legend('$$E^Q\Big[\frac{S(t)}{M(t)}\Big]$$','paths $$\frac{S(t)}{M(t)}$$');
set(leg,'Interpreter','latex');

plot(time,S_0*ones(length(time),1),'--r','linewidth',2)
titleT=title('Monte Carlo Paths of $$\frac{S(t)}{M(t)}$$ under risk-neutral measure, Q');
set(titleT,'Interpreter','latex');
